src/coreComponents/linearAlgebra/interfaces/direct/Arnoldi.hpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 Total, S.A
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file Arnoldi.hpp
+ */
+
+#ifndef GEOSX_LINEARALGEBRA_SOLVERS_ARNOLDI_HPP_
+#define GEOSX_LINEARALGEBRA_SOLVERS_ARNOLDI_HPP_
+
+#include "linearAlgebra/interfaces/VectorBase.hpp"
+#include "linearAlgebra/interfaces/BlasLapackLA.hpp"
+
+namespace geosx
+{
+
+/**
+ * @brief DirectOperator Simple class to apply the operator A^T * A to a vector
+ */
+template< typename MATRIX, typename VECTOR >
+class DirectOperator
+{
+public:
+
+  /**
+   * @brief Sets the matrix
+   * @param matrix the matrix
+   */
+  void set( MATRIX const & matrix )
+  {
+    m_matrix = &matrix;
+    m_comm = matrix.getComm();
+  }
+
+  /**
+   * @brief Returns the global size
+   * @return the global size
+   */
+  globalIndex globalSize() const
+  {
+    return m_matrix->numGlobalRows();
+  }
+
+  /**
+   * @brief Returns the local size
+   * @return the local size
+   */
+  localIndex localSize() const
+  {
+    return m_matrix->numLocalRows();
+  }
+
+  /**
+   * @brief Returns the communicator
+   * @return the communicator
+   */
+  MPI_Comm const & getComm() const
+  {
+    return m_comm;
```

Comment:
For example in `ArnoldiLargestEigenvalue` only the _operator_ is known, so it has to contain the communicator info.
https://github.com/GEOSX/GEOSX/blob/e40aa066a520c4b5dce9fe9382e43ca0057bb2ca/src/coreComponents/linearAlgebra/interfaces/direct/Arnoldi.hpp#L100

Personally, I don't like the idea of `op.getMatrix().getComm()`. The operator is more generic than a matrix-vector product ...

Reply:
Sure, I'd totally support making it more generic and being able to handle arbitrary abstract operators. At the moment though, you're calling `matrix.getComm()` in `DirectOperator::set()`, which will only work with an operator type that has that method, and my comment was that you could just do the same in this function instead of saving `m_comm`. Alternatively you might pass the communicator as a separate argument in `set()`, so it would be upon the caller to provide it correctly. This is all just minor nitpicking and not super important for this PR to be merged.

In my head it raises the question whether we want things like communicator and parallel partitioning (local rows/cols) associated with "abstract" operators rather than "concrete" matrix classes, but perhaps it's another discussion.

Reply:
Ok, fine! I got your point and I agree. Now the `set` method requires two arguments: matrix and communicator.